### PR TITLE
ValuesSourceRegistry Prototype

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR builds out a prototype implementation of the `ValuesSourceRegistry` to explore the API.  As an example, I've done some of the conversion work on `HistogramAggregatorFactory` to show how this can work.  We may need to update this API for more complex aggregations like Terms family.

Related to #42949

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregatorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.support;
+
+public interface AggregatorSupplier {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregatorSupplier.java]: I don't love having a no-op interface here, but it gives us a bit of type safety over just passing in any old object.  The problem here is that each aggregation family has a unique signature for how it constructs aggregators, so we can't have a common interface that means something.  This gives us a type that all those aggregation specific suppliers can inherit from, and we can be reasonably sure that it's a safe cast, since we trust that people aren't cross-registering aggregators.  And, if there were a cross-registered aggregator, it wouldn't work anyway, so failing if we detect a bad cast is ok too.

That said, this feels really clunky and I'm open to better options.  I spent a lot of time thinking about this problem, and this was the best solution I came up with.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/HistogramAggregatorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public interface HistogramAggregatorSupplier extends AggregatorSupplier {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/HistogramAggregatorSupplier.java]: Each aggregation family will define one of these which wraps the constructor signature for the aggregators that go with that family.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 3]
```diff
@@ -68,15 +68,22 @@
...
+                throw new IllegalStateException(
+                    "value source config is invalid; must have either a field context or a script or marked as unwrapped");
             }
+            /*
+             * This is the Stand Alone Script path.  We should have a script that will produce a value independent of the presence or
+             * absence of any one field.  The type of the script is given by the valueType field, and defaults to bytes if not specified.
+             */
+            // TODO: Not pluggable, should always be valueType if specified, BYTES if not.
+            // TODO: Probably should validate that the resulting type is valid for this agg.  That needs to be plugable.
             ValuesSourceType valuesSourceType = valueType != null ? valueType.getValuesSourceType() : ValuesSourceType.ANY;
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Worth noting that by the time we get here, `valueType` is hopelessly muddled.  It could be user-specified, or it could be a default value from the aggregation builder, or it could be null.  We have no way of knowing the source at this point.  I have a plan to clean that up, but in the interest of small, manageable PRs, I want to do that in another PR.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java [hunk 4]
```diff
@@ -85,53 +92,45 @@
...
+        } else {
+            // Field case
+            MappedFieldType fieldType = context.fieldMapper(field);
+            if (fieldType == null) {
+                /* Unmapped Field Case
+                 * We got here because the user specified a field, but it doesn't exist on this index, possibly because of a wildcard index
+                 * pattern.  In this case, we're going to end up using the EMPTY variant of the ValuesSource, and possibly applying a user
+                 * specified missing value.
+                 */
+                // TODO: This should be pluggable too; Effectively that will replace the missingAny() case from toValuesSource()
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Again, this TODO will happen in another PR centered around fixing ValueType

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+/*
+This is a _very_ crude prototype for the ValuesSourceRegistry which basically hard-codes everything.  The intent is to define the API
+for aggregations using the registry to resolve aggregators.
+ */
+public enum ValuesSourceRegistry {
+    INSTANCE {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java]: Sorry about the funky mutable enum singleton pattern, I just needed something that wasn't going to get in the way for prototyping.  Happy to use a different implementation before going to production. 

This is intended to illustrate the core API for the new registry pattern.  We'll probably need to add one or two more registered components, for the script and unmapped branches in `ValuesSourceConfig.resolve()`, but in general this is how it should work: for each aggregation family, a mapping of predicates to `ValuesSourceType`s and a mapping of `ValuesSourceType`s to `AggregatorSuppliers` (where the actual aggregator supplier implements an aggregation family specific sub-interface).  

This is the heart of the plan, so please raise any concerns with this, no matter how trivial.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/NumericHistogramAggregatorTests.java [hunk 6]
```diff
@@ -100,6 +106,41 @@ public void testDoubles() throws Exception {
         }
     }
 
+    public void testDates() throws Exception {
```
Comment (not-napoleon) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/NumericHistogramAggregatorTests.java]: I'm not sure we should support this, but we do currently, so I wanted a test for the legacy case.

**Review by polyfractal — APPROVED** [review]
I think this LGTM.  All the comments I had while looking over it are already noted in the code or with TODOs, and it's just an iterative building block too.  :+1:
